text/0000-date-api.md

```diff
@@ -0,0 +1,60 @@
+- Feature Name: date_api
+- Start Date: 2018-03-21
+- RFC PR: 
+- Neon Issue: 
+
+# Summary
+[summary]: #summary
+
+This proposal defines an API for creating and inspecting JavaScript `Date` objects.
+
+# Motivation
+[motivation]: #motivation
+
+The `Date` type is part of the standard JavaScript library and should be exposed to Rust.
+
+Example:
+
+```rust
+let date: Handle<JsDate> = JsDate::new(1431673200000)?;
+let value: f64 = date.value(); // 1431673200000.0
+```
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+This is a relatively straightforward API and can be explained with the API docs.
+
+# Details
+[details]: #details
+
+## `JsDate`
+
+This API adds a `JsDate` type to `neon::types`:
+
+```rust
+impl JsDate {
+    pub fn new<'a, S: Scope<'a>, V: Into<f64>>(_: &mut S, value: V) -> JsResult<JsDate>;
```

Comment:
Some `f64` are invalid dates. Should we immediately throw or should we provide an intermediary error that allows the user to handle or call `.or_throw` the user to use `.or_throw`?

Reply:
Good question. Constructing an invalid date is not an error in JS:

```text
> var d = new Date(Number.MAX_SAFE_INTEGER + 1000)
undefined
> d
Invalid Date
```

So I think it shouldn't be an error, in order to fully model the semantics of JS.

Reply:
Maybe we could also add an `is_valid(&self) -> bool` method.

Reply:
The range of [legal date values](https://www.ecma-international.org/ecma-262/11.0/index.html#sec-time-values-and-time-range) is `[-8640000000000000, 8640000000000000]`:

```
> new Date(-8640000000000001)
Invalid Date
> new Date(-8640000000000000)
-271821-04-20T00:00:00.000Z
> new Date(8640000000000000)
+275760-09-13T00:00:00.000Z
> new Date(8640000000000001)
Invalid Date
```

Reply:
I added an `is_valid()` method and associated constants for the min and max legal date values.